ssl/ech/ech_internal.c

```diff
@@ -160,4 +214,1194 @@ int ossl_ech_conn_init(SSL_CONNECTION *s, SSL_CTX *ctx,
+    BUF_MEM *tpkt_mem = NULL, *shpkt_mem = NULL;
+
+    /*
+     * SH preamble has bad length at this point on server
+     * and is missing on client so we'll fix
+     */
+    if ((shpkt_mem = BUF_MEM_new()) == NULL
+        || !WPACKET_init(&shpkt, shpkt_mem)) {
+        BUF_MEM_free(shpkt_mem);
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if (!WPACKET_put_bytes_u8(&shpkt, SSL3_MT_SERVER_HELLO)
+        || (s->server == 1 && !WPACKET_put_bytes_u24(&shpkt, shlen - 4))
+        || (s->server == 1 && !WPACKET_memcpy(&shpkt, shbuf + 4, shlen -4))
+        || (s->server == 0 && !WPACKET_put_bytes_u24(&shpkt, shlen))
+        || (s->server == 0 && !WPACKET_memcpy(&shpkt, shbuf, shlen))
+        || !WPACKET_get_length(&shpkt, fixedshbuf_len)) {
+        BUF_MEM_free(shpkt_mem);
+        WPACKET_cleanup(&shpkt);
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    fixedshbuf = (unsigned char *)shpkt_mem->data;
+    WPACKET_cleanup(&shpkt);
+# ifdef OSSL_ECH_SUPERVERBOSE
+    ossl_ech_pbuf("cx: fixed sh buf", fixedshbuf, *fixedshbuf_len);
+# endif
+    if ((tpkt_mem = BUF_MEM_new()) == NULL
+        || !WPACKET_init(&tpkt, tpkt_mem)) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if (s->hello_retry_request == SSL_HRR_NONE) {
+        if (!WPACKET_memcpy(&tpkt, s->ext.ech.innerch,
+                            s->ext.ech.innerch_len)
+            || !WPACKET_get_length(&tpkt, chend)
+            || !WPACKET_memcpy(&tpkt, fixedshbuf, *fixedshbuf_len)
+            || !WPACKET_get_length(&tpkt, tlen)) {
+            SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+        *tbuf = OPENSSL_malloc(*tlen);
+        if (*tbuf == NULL) {
+            SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+        memcpy(*tbuf, WPACKET_get_curr(&tpkt) - *tlen, *tlen);
+        BUF_MEM_free(shpkt_mem);
+        WPACKET_cleanup(&tpkt);
+        BUF_MEM_free(tpkt_mem);
+        return 1;
+    }
+    /* everything below only applies if we're at some stage in doing HRR */
+    if ((md = (EVP_MD *)ssl_handshake_md(s)) == NULL) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    if (for_hrr == 0) { /* after 2nd SH rx'd */
+        hashin = s->ext.ech.innerch1;
+        hashin_len = s->ext.ech.innerch1_len;
+    } else { /* after HRR rx'd */
+        hashin = s->ext.ech.innerch;
+        hashin_len = s->ext.ech.innerch_len;
+        OPENSSL_free(s->ext.ech.kepthrr);
+        /* stash this SH/HRR for later */
+        s->ext.ech.kepthrr = OPENSSL_malloc(*fixedshbuf_len);
+        if (s->ext.ech.kepthrr == NULL) {
+            SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+        memcpy(s->ext.ech.kepthrr, fixedshbuf, *fixedshbuf_len);
```

Comment:
Why not just:
```suggestion
        s->ext.ech.kepthrr = fixedshbuf;
```

Also...with the code as it currently stands I don't see where `fixedshbuf` gets freed?

Reply:
OBE - see below